File: swift/DesignPatterns/Creational/FactoryMethod.swift
//
//  FactoryMethod.swift
//  DesignPatterns
//
//  Created by Soe Min Thein on 28/05/2026.
//

import Foundation

protocol Shape {
    func draw()
}

struct Circle: Shape {
    func draw() {
        print("Drawing a circle")
    }
}

struct Rectangle: Shape {
    func draw() {
        print("Drawing a rectangle")
    }
}

enum ShapeType {
    case circle
    case rectangle
}

struct ShapeFactory {
    static func makeShape(_ type: ShapeType) -> Shape {
        switch type {
        case .circle:
            return Circle()
        case .rectangle:
            return Rectangle()
        }
    }
}

func factoryExample() {
    let circle = ShapeFactory.makeShape(.circle)
    circle.draw()
    
    let rectangle = ShapeFactory.makeShape(.rectangle)
    rectangle.draw()
}

File: dart/design_patterns/lib/creational/factory_method.dart
// Product

abstract class Shape {
  void draw();
}

// Concrete Products

class Circle implements Shape {
  @override
  void draw() {
    print('Drawing a circle');
  }
}

class Rectangle implements Shape {
  @override
  void draw() {
    print('Drawing a rectangle');
  }
}

// Shape Types

enum ShapeType { circle, rectangle }

// Factory

class ShapeFactory {
  static Shape makeShape(ShapeType type) {
    switch (type) {
      case ShapeType.circle:
        return Circle();
      case ShapeType.rectangle:
        return Rectangle();
    }
  }
}

// Usage

void main() {
  final circle = ShapeFactory.makeShape(ShapeType.circle);
  circle.draw();

  final rectangle = ShapeFactory.makeShape(ShapeType.rectangle);
  rectangle.draw();
}
